velox/functions/prestosql/benchmarks/ArraySumBenchmark.cpp

```diff
@@ -0,0 +1,213 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <folly/Benchmark.h>
+#include <folly/init/Init.h>
+#include "velox/expression/VectorFunction.h"
+#include "velox/functions/Macros.h"
+#include "velox/functions/Registerer.h"
+#include "velox/functions/lib/LambdaFunctionUtil.h"
+#include "velox/functions/lib/benchmarks/FunctionBenchmarkBase.h"
+#include "velox/functions/prestosql/registration/RegistrationFunctions.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::exec;
+using namespace facebook::velox::functions;
+
+namespace {
+
+///
+/// Implements the array_sum function.
+/// See documentation at https://prestodb.io/docs/current/functions/array.html
+///
+template <typename IT, typename OT>
+class ArraySumFunction : public exec::VectorFunction {
+ public:
+  void apply(
+      const SelectivityVector& rows,
+      std::vector<VectorPtr>& args, // Not using const ref so we can reuse args
+      const TypePtr& outputType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const override {
+    // Acquire the array elements vector.
+    auto arrayVector = args[0]->as<ArrayVector>();
+    VELOX_CHECK(arrayVector);
+    auto elementsVector = arrayVector->elements();
+    auto elementsRows =
+        toElementRows(elementsVector->size(), rows, arrayVector);
+    exec::LocalDecodedVector elements(context, *elementsVector, elementsRows);
+    vector_size_t numRows = arrayVector->size();
+
+    // Prepare result vector for writing
+    BaseVector::ensureWritable(rows, outputType, context->pool(), result);
+    auto resultValues = (*result)->template asFlatVector<OT>();
+
+    rows.template applyToSelected([&](vector_size_t row) {
```

Comment:
For the sake of benchmarking, it is necessary to add fast paths for flat no-nulls cases. Otherwise, there is no point in comparing.

Reply:
Could you point me to somewhere in the code this is done?

Reply:
Here is an example: https://github.com/facebookincubator/velox/pull/1904/files#diff-9a136877eecc20521d0690bb3c5942ce187ddd0e72f0abe2a3d47e601e8348c5R37